crates/sui-light-client/src/main.rs

```diff
@@ -0,0 +1,714 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::anyhow;
+use async_trait::async_trait;
+use move_core_types::account_address::AccountAddress;
+use sui_json_rpc_types::SuiTransactionBlockResponseOptions;
+
+use sui_rest_api::{CheckpointData, Client};
+use sui_types::{
+    base_types::{ObjectID, SequenceNumber},
+    committee::Committee,
+    crypto::AuthorityQuorumSignInfo,
+    digests::TransactionDigest,
+    effects::{TransactionEffects, TransactionEffectsAPI, TransactionEvents},
+    message_envelope::Envelope,
+    messages_checkpoint::{CertifiedCheckpointSummary, CheckpointSummary, EndOfEpochData},
+    object::Data,
+};
+
+use sui_config::genesis::Genesis;
+
+use sui_json::SuiJsonValue;
+use sui_package_resolver::Result as ResolverResult;
+use sui_package_resolver::{Package, PackageStore, Resolver};
+use sui_sdk::SuiClientBuilder;
+
+use clap::{Parser, Subcommand};
+use std::{
+    fs,
+    io::Write,
+    path::{Path, PathBuf},
+    str::FromStr,
+};
+use std::{io::Read, sync::Arc};
+
+/// A light client for the Sui blockchain
+#[derive(Parser, Debug)]
+#[command(author, version, about, long_about = None)]
+struct Args {
+    /// Sets a custom config file
+    #[arg(short, long, value_name = "FILE")]
+    config: Option<PathBuf>,
+
+    #[command(subcommand)]
+    command: Option<SCommands>,
+}
+
+struct RemotePackageStore {
+    client: Client,
+    config: Config,
+}
+
+impl RemotePackageStore {
+    pub fn new(client: Client, config: Config) -> Self {
+        Self { client, config }
+    }
+}
+
+#[async_trait]
+impl PackageStore for RemotePackageStore {
+    /// Latest version of the object at `id`.
+    async fn version(&self, id: AccountAddress) -> ResolverResult<SequenceNumber> {
+        Ok(self.client.get_object(id.into()).await.unwrap().version())
+    }
+    /// Read package contents. Fails if `id` is not an object, not a package, or is malformed in
+    /// some way.
+    async fn fetch(&self, id: AccountAddress) -> ResolverResult<Arc<Package>> {
+        let object = self.client.get_object(id.into()).await.unwrap();
+
+        // Need to authenticate this object
+        let (effects, _) =
+            get_verified_effects_and_events(&self.config, object.previous_transaction)
+                .await
+                .unwrap();
+        // check that this object ID, version and hash is in the effects
+        effects
+            .all_changed_objects()
+            .iter()
+            .find(|object_ref| object_ref.0 == object.compute_object_reference())
+            .unwrap();
```

Comment:
Seems redundant logic to when we authenticate objects below

Reply:
You are right, I now unified the logic under a single function `get_verified_object`.